docs/development/coding-style-guide.md

```diff
@@ -0,0 +1,262 @@
+# Lokomotive coding style guide
+
+This document describes the coding style that's enforced in the Lokomotive codebase. The guidelines
+in this document should be followed by Lokomotive maintainers and contributors alike. The guide is
+structured in a way which makes it easy to link to a specific guideline, for example from a PR
+comment.
+
+## Guidelines
+
+This section describes the entire set of coding style guidelines which should be followed when
+working on the Lokomotive codebase.
+
+Each "topic" should reside on its own under a heading so that it's easy to unambiguously link to a
+specific guideline during code reviews.
+
+The list of guidelines is meant to evolve over time. If you think a new guideline should be
+enforced, suggest changes to this document in a PR.
+
+### Error string formatting
+
+Error strings are typically *chained*:
+
+```
+Deploying cluster: reading component config: parsing HCL: unexpected character '%'
+```
+
+Remember this fact when constructing error strings and think about how the final string might look
+to the user.
+
+#### Wrapping errors
+
+When "wrapping" an error in another error to provide context, the `%w` formatting verb should be
+used following the [Go 1.13 recommendations](https://blog.golang.org/go1.13-errors) to error
+handling:
+
+`fmt.Errorf("parsing config: %w", err)`
+
+#### Error string text
+
+Don't include words which describe failure in error strings. This includes words such as "failed",
+"error", "could not" etc. It is clear that an error indicates a failure. Instead, provide just the
+*context*, i.e. **what we were doing** when the error occurred (or the erroneous condition if the
+error is the last one in the "chain").
+
+Good:
+
+- `fmt.Errorf("parsing config: %w", err)`
+- `fmt.Errorf("deploying cluster: %w", err)`
+- `fmt.Errorf("reversing time: %w", err)`
```

Comment:
I don't really know what that is. In fact, the "bad" alternative was more useful for me as at least there I'm slightly more confident that to "reverse time" is a thing.

Maybe this fact signals that error messages like this are a little cryptic (at least if you are not familiar with the concepts, as I guess several users won't be familiar with what HCL is)?

Reply:
@rata where is HCL mentioned here?

I think the main motivation for removing `failed` and similar keywords is to avoid cluttering the code with them, as you would need to include them in every single error message. And `fmt.Errorf` already implies that something has failed.

Perhaps to provide some extra context here, it would be good to have an example how chained errors looks like later on. e.g.
`Applying cluster failed: parsing configuration: illegal character at line 8`.
vs
`Applying cluster failed: parsing configuration failed: failed to parse illegal character at line 8`.

Maybe it would make more sense to do something like:
`Applying cluster failed. Cause: parsing configuration: illegal character at line 8`.

BTW, now that I think about it, perhaps we should also write something down, that those error traces should be uniquely identifiable, so ideally single function shouldn't return the same error in 5 different places, as then it makes it difficult to trace what exactly happened.

Reply:
@rata "reversing time" was included as an example only. I'd be extremely careful with adding actual time reversal functionality to Lokomotive, and in any case this should be done in a major release.

The intention behind the error formatting guidelines is to avoid having things like the following:

`Deploying cluster: Could not deploy cluster: Failed to deploy cluster due to an error: Error while running Terraform: Terraform execution failed: Error - no such file or directory`

Reply:
@invidian I don't think that makes much sense. It's a log, not a core dump :-D

@johananl Ohhh, I get the time reverse now! :joy:. I got the intention to avoid those kind of errors, was unsure if the whole error was better one way or the other. But I guess this seems fine. Thanks :)

Reply:
I've made it clearer why we shouldn't "clutter" error strings with "failed", "couldn't" etc.